processing/src/main/java/io/druid/query/groupby/RowBasedColumnSelectorFactory.java

```diff
@@ -228,7 +229,7 @@ public boolean matches()
                 }
 
                 for (String dimensionValue : dimensionValues) {
-                  if (Objects.equals(Strings.emptyToNull(dimensionValue), value)) {
+                  if (Objects.equals(NullHandling.emptyToNullIfNeeded(dimensionValue), value)) {
```

Comment:
Group By experts - @gianm / @jon-wei / @jihoonson - could you please check that this change is valid (same below)

Reply:
Also, a more general question -- how do you distinguish between empty strings that come from "old" segment format (and then should be converted to nulls), and empty strings that come from "new" segment format, and then empty string is a "proper" empty string, and shouldn't be converted?

Shouldn't use of `emptyToNullIfNeeded()` be localized *only* to `GenericIndexed` that handles old segment format?

Reply:
I agree on the behavior of old data that empty strings should be converted to null when the segment is read, but that is not the case at present since full query layer is not capable of handling null values and at some places like filters we specifically match for empty strings. 
Note this PR only has partial changes to support new storage format and not touches the older format and query layer to make it short.

Reply:
I think the `NullHandling.emptyToNullIfNeeded` calls may be unnecessary here, since I believe this value matcher would only be used when filtering the result of a groupBy subquery, so the subquery results should have already applied the necessary conversions when running in the older "empty == null" mode

Reply:
tried removing emptyToNull call, but that lead to Unit test failures for filters, in BaseFilterTest at selectColumnValuesMatchingFilterUsingRowBasedColumnSelectorFactory which specifically tries to test this, So it seems we may need this for the case tested there. 
since this PR does not introduce this as a new change in RowBasedColumnSelectorFactory, it may be ok to keep the check as it is.